### PR TITLE
qutebrowser: support for webrtc pipewire (needed for wayland native screen sharing support)

### PR DESCRIPTION
**Pull request by johnae**
###### Motivation for this change

Qutebrowser can support screen sharing natively on Wayland but it requires a QtWebEngine built with such support as well as making the pipewire library available to Qutebrowser at runtime. This pull-request does that.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by johnae** [comment]
Importantly qutebrowser must be started with  `--qt-flag enable-features=WebRTCPipeWireCapturer` for this to actually work.

File: pkgs/development/libraries/qt-5/modules/qtwebengine.nix [hunk 1]
```diff
@@ -140,7 +141,7 @@ qtModule {
     fi
   '';
 
-  qmakeFlags = [ "--" "-system-ffmpeg" ]
+  qmakeFlags = [ "--" "-system-ffmpeg" "-webengine-webrtc-pipewire"]
```
Comment (veprbl) [inline pkgs/development/libraries/qt-5/modules/qtwebengine.nix:144]:
```suggestion
  qmakeFlags = [ "--" "-system-ffmpeg" ]
    ++ optional (!stdenv.isDarwin) "-webengine-webrtc-pipewire"
```

File: pkgs/applications/networking/browsers/qutebrowser/default.nix [hunk 2]
```diff
@@ -112,12 +112,18 @@ in mkDerivationWith python3Packages.buildPythonApplication rec {
...
+  preFixup = let
+    libPath = lib.makeLibraryPath [ pipewire_0_2 ];
+  in
+    ''
     makeWrapperArgs+=(
       "''${gappsWrapperArgs[@]}"
       "''${qtWrapperArgs[@]}"
       --add-flags '--backend ${backend}'
       --set QUTE_QTWEBENGINE_VERSION_OVERRIDE "${lib.getVersion qtwebengine}"
+      ${if backend == "webengine" then
```
Comment (veprbl) [inline pkgs/applications/networking/browsers/qutebrowser/default.nix:124]:
```suggestion
      ${lib.optionalString (!stdenv.isDarwin) && (backend == "webengine") ''--prefix LD_LIBRARY_PATH : ${libPath}''}
```

Comment (johnae) [reply pkgs/applications/networking/browsers/qutebrowser/default.nix]: Fixed.

File: pkgs/development/libraries/qt-5/modules/qtwebengine.nix (hunk 1 above)
Comment (johnae) [reply pkgs/development/libraries/qt-5/modules/qtwebengine.nix]: Fixed.

**Comment by johnae** [comment]
Thank you for the review @veprbl!

**Comment by johnae** [comment]
On sway you need xdg-desktop-portal-wlr or it won't work. Also, while
Google meet supposedly works in QTWebEngine (needs hangouts extension which
should be part of 5.15.3) I can't get it to work.

I've successfully tested screen sharing on Mozillas gum test page here:

https://mozilla.github.io/webrtc-landing/gum_test.html


Den tors 20 maj 2021 15:53Michele Guerini Rocco ***@***.***>
skrev:

**Comment by johnae** [comment]
You need something like this:

```nix
services.pipewire.enable = true;
xdg.portal.enable = true;
xdg.portal.gtkUsePortal = true;
xdg.portal.extraPortals = [ pkgs.xdg-desktop-portal-wlr
pkgs.xdg-desktop-portal-gtk ];
```

Den tors 20 maj 2021 17:33John Axel Eriksson ***@***.***> skrev:

**Comment by johnae** [comment]
You need something like this:

```nix
services.pipewire.enable = true;
xdg.portal.enable = true;
xdg.portal.gtkUsePortal = true;
xdg.portal.extraPortals = [ pkgs.xdg-desktop-portal-wlr
pkgs.xdg-desktop-portal-gtk ];
```

**Comment by johnae** [comment]
For reference, here's where Google Meet (which needs the hangouts extension apparently) is discussed on QT:s bugtracker: https://bugreports.qt.io/browse/QTBUG-85731?focusedCommentId=530973

**Comment by rnhmjoj** [comment]
> You need something like this:
> 
> ```nix
> services.pipewire.enable = true;
> xdg.portal.enable = true;
> xdg.portal.gtkUsePortal = true;
> xdg.portal.extraPortals = [ pkgs.xdg-desktop-portal-wlr pkgs.xdg-desktop-portal-gtk ];
> ```

Thank you, this worked. Btw, only `xdg-desktop-portal-gtk` is needed:
the other is already included by the sway module.

Screen sharing is working fine, now.

**Review by rnhmjoj — APPROVED** [review]

**Comment by flokli** [comment]
The QT changes broke the `qt514.qtwebengine` build, as that version of Qt doesn't seem to know about this configure flag, and also increases the build closure unnecessarily. Fix in https://github.com/NixOS/nixpkgs/pull/124078.

**Pull request merged**